### PR TITLE
More robust dt_import_session_path checks

### LINKED ISSUE #8938: Image tethering : picture import segfaults
While taking pictures with Nikon Z6 in tethered mode, everything works until import. Then it crashes with:

1. if `codepath/openmp_simd = true`:

```
0x00007ffff2f46afe in __strcmp_avx2 () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff2f46afe in __strcmp_avx2 () at /lib64/libc.so.6
#1  0x00007ffff6786f87 in gp_widget_get_child_by_name () at /lib64/libgphoto2.so.6
#2  0x00007ffff6786fb1 in gp_widget_get_child_by_name () at /lib64/libgphoto2.so.6
#3  0x00007ffff7e337c8 in dt_camctl_camera_property_exists (c=<optimized out>, cam=0x7fff4803fdc0, cam@entry=0x0, property_name=property_name@entry=0x7fff42765745 "expprogram")
    at /home/aurelienpierre/Documents/Programmes/darktable/src/common/camera_control.c:1599
#4  0x00007fff42765018 in _lib_property_add_new (propertyname=0x7fff42765745 "expprogram", label=0x7fff4276573d "program", lib=0x4f2abd0)
    at /home/aurelienpierre/Documents/Programmes/darktable/src/libs/camera.c:126
#5  view_enter (self=<optimized out>, old_view=<optimized out>, new_view=<optimized out>) at /home/aurelienpierre/Documents/Programmes/darktable/src/libs/camera.c:568
#6  0x00007ffff7e27dac in dt_view_manager_switch_by_view (vm=0x3e60290, nv=0x3ebb560) at /home/aurelienpierre/Documents/Programmes/darktable/src/views/view.c:395
#7  0x00007ffff7e28139 in dt_view_manager_switch (vm=<optimized out>, view_name=view_name@entry=0x7fff642d7968 "tethering")
    at /home/aurelienpierre/Documents/Programmes/darktable/src/views/view.c:231
#8  0x00007ffff7c5fc60 in _dt_ctl_switch_mode_to (user_data=0x7fff642d7968) at /home/aurelienpierre/Documents/Programmes/darktable/src/control/control.c:356
#9  0x00007ffff780e2e5 in g_main_context_invoke_full () at /lib64/libglib-2.0.so.0
#10 0x00007ffff6a58000 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#11 0x00007ffff6a58123 in g_signal_emit () at /lib64/libgobject-2.0.so.0
#12 0x00007ffff714a120 in gtk_real_button_released () at /lib64/libgtk-3.so.0
#13 0x00007ffff6a58000 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#14 0x00007ffff6a58123 in g_signal_emit () at /lib64/libgobject-2.0.so.0
#15 0x00007ffff7149ca4 in multipress_released_cb () at /lib64/libgtk-3.so.0
#16 0x00007ffff740252f in _gtk_marshal_VOID__INT_DOUBLE_DOUBLEv () at /lib64/libgtk-3.so.0
#17 0x00007ffff6a58000 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#18 0x00007ffff6a58123 in g_signal_emit () at /lib64/libgobject-2.0.so.0
#19 0x00007ffff7216a8c in gtk_gesture_multi_press_end () at /lib64/libgtk-3.so.0
#20 0x00007ffff6a3f014 in g_cclosure_marshal_VOID__BOXEDv () at /lib64/libgobject-2.0.so.0
#21 0x00007ffff6a58000 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#22 0x00007ffff6a58123 in g_signal_emit () at /lib64/libgobject-2.0.so.0
#23 0x00007ffff720fd5b in _gtk_gesture_check_recognized.lto_priv.0 () at /lib64/libgtk-3.so.0
#24 0x00007ffff72180a3 in gtk_gesture_handle_event () at /lib64/libgtk-3.so.0
#25 0x00007ffff7218a03 in gtk_gesture_single_handle_event () at /lib64/libgtk-3.so.0
#26 0x00007ffff71d5500 in gtk_event_controller_handle_event () at /lib64/libgtk-3.so.0
#27 0x00007ffff73b50e5 in _gtk_widget_run_controllers.lto_priv.0 () at /lib64/libgtk-3.so.0
#28 0x00007ffff7400bec in _gtk_marshal_BOOLEAN__BOXEDv () at /lib64/libgtk-3.so.0
#29 0x00007ffff6a58000 in g_signal_emit_valist () at /lib64/libgobject-2.0.so.0
#30 0x00007ffff6a58123 in g_signal_emit () at /lib64/libgobject-2.0.so.0
#31 0x00007ffff73c3ad4 in gtk_widget_event_internal.part.0.lto_priv () at /lib64/libgtk-3.so.0
#32 0x00007ffff7261d20 in propagate_event.lto_priv () at /lib64/libgtk-3.so.0
#33 0x00007ffff7262ff3 in gtk_main_do_event () at /lib64/libgtk-3.so.0
#34 0x00007ffff6f3c673 in _gdk_event_emit () at /lib64/libgdk-3.so.0
#35 0x00007ffff6f744f6 in gdk_event_source_dispatch () at /lib64/libgdk-3.so.0
#36 0x00007ffff780ea9f in g_main_context_dispatch () at /lib64/libglib-2.0.so.0
#37 0x00007ffff7860a98 in g_main_context_iterate.constprop () at /lib64/libglib-2.0.so.0
#38 0x00007ffff780e163 in g_main_loop_run () at /lib64/libglib-2.0.so.0
#39 0x00007ffff726048d in gtk_main () at /lib64/libgtk-3.so.0
#40 0x00007ffff7dc8879 in dt_gui_gtk_run (gui=<optimized out>) at /home/aurelienpierre/Documents/Programmes/darktable/src/gui/gtk.c:1564
#41 0x0000000000400970 in main (argc=<optimized out>, argv=<optimized out>) at /home/aurelienpierre/Documents/Programmes/darktable/src/main.c:93
```
2. else:
```
0x00007ffff7be1687 in _image_import_internal (film_id=<optimized out>, filename=<optimized out>, override_ignore_jpegs=override_ignore_jpegs@entry=1, 
    raise_signals=raise_signals@entry=1, lua_locking=1) at /home/aurelienpierre/Documents/Programmes/darktable/src/common/image.c:1600
1600      img->group_id = group_id;
(gdb) bt
#0  0x00007ffff7be1687 in _image_import_internal
    (film_id=<optimized out>, filename=<optimized out>, override_ignore_jpegs=override_ignore_jpegs@entry=1, raise_signals=raise_signals@entry=1, lua_locking=1)
    at /home/aurelienpierre/Documents/Programmes/darktable/src/common/image.c:1600
#1  0x00007ffff7be2455 in dt_image_import
    (film_id=<optimized out>, filename=<optimized out>, override_ignore_jpegs=override_ignore_jpegs@entry=1, raise_signals=raise_signals@entry=1)
    at /home/aurelienpierre/Documents/Programmes/darktable/src/common/image.c:1695
#2  0x00007ffff7c6cbbe in dt_image_import_job_run (job=0x7fff3c108510) at /home/aurelienpierre/Documents/Programmes/darktable/src/control/jobs/image_jobs.c:79
#3  0x00007ffff7c63d26 in dt_control_job_execute (job=job@entry=0x7fff3c108510) at /home/aurelienpierre/Documents/Programmes/darktable/src/control/jobs.c:300
#4  0x00007ffff7c64828 in dt_control_run_job (control=0x443dc0) at /home/aurelienpierre/Documents/Programmes/darktable/src/control/jobs.c:319
#5  dt_control_work (ptr=<optimized out>) at /home/aurelienpierre/Documents/Programmes/darktable/src/control/jobs.c:564
#6  0x00007ffff7a7d3f9 in start_thread () at /lib64/libpthread.so.0
#7  0x00007ffff2eeab53 in clone () at /lib64/libc.so.6

```

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
checks for a writable existing directory when finding the path. Also checks for that before trying to
keep the current session path as that might be not ok any more.

Likely Fixes #8938

As i don't use import sessions often i might have overseen something in my tests, please check and confirm.

### CODE REVIEW

File: src/common/import_session.c [hunk 1]
```diff
@@ -307,14 +307,36 @@ const char *dt_import_session_filename(struct dt_import_session_t *self, gboolea
   return self->current_filename;
 }
 
+gboolean _dt_test_writable_dir(const char *path)
```
Comment (TurboGit) [inline src/common/import_session.c:310]: I would move that into utility.h/c next to dt_util_test_image_file().

**Review by TurboGit — CHANGES_REQUESTED** [review]
Minor refactoring.

File: src/common/import_session.c (hunk 1 above)
Comment (jenshannoschwalm) [reply src/common/import_session.c]: good one.

File: src/common/utility.c [hunk 2]
```diff
@@ -294,6 +294,22 @@ gboolean dt_util_test_image_file(const char *filename)
...
+  if(path == NULL) return FALSE;
+#ifdef _WIN32
+  struct _stati64 stats;
+  if(_stati64(path, &stats)) return FALSE;
+#else
+  struct stat stats;
+  if(stat(path, &stats)) return FALSE;
+#endif
+  if(S_ISDIR(stats.st_mode) == 0) return FALSE;  
+  if(g_access(path, W_OK | X_OK) != 0) return FALSE;
```
Comment (TurboGit) [inline src/common/utility.c:308]: Can't we avoid a call to g_access() here by check the stat st_mod filed. I quote the Linux man:


```
       The following mask values are defined for the file  mode  component  of
       the st_mode field:

           S_ISUID     04000   set-user-ID bit (see execve(2))
           S_ISGID     02000   set-group-ID bit (see below)
           S_ISVTX     01000   sticky bit (see below)

           S_IRWXU     00700   owner has read, write, and execute permission
           S_IRUSR     00400   owner has read permission
           S_IWUSR     00200   owner has write permission
           S_IXUSR     00100   owner has execute permission

           S_IRWXG     00070   group has read, write, and execute permission
           S_IRGRP     00040   group has read permission
           S_IWGRP     00020   group has write permission
           S_IXGRP     00010   group has execute permission

           S_IRWXO     00007   others  (not  in group) have read, write, and
                               execute permission
           S_IROTH     00004   others have read permission
           S_IWOTH     00002   others have write permission
           S_IXOTH     00001   others have execute permission
```

Comment (jenshannoschwalm) [reply src/common/utility.c:308]: I think that's exactly what @cryptomilk meant. For linux it's that easy, sure. But for windows, i am not so sure. Reading through the docs for g_stat and g_access and references didn't help here, one further reference

https://web.archive.org/web/20101214222704/http://msdn.microsoft.com/en-us/library/14h5k7ff(v=vs.80).aspx

makes me very suspicious about it and so i trusted g_access.

Comment (TurboGit) [reply src/common/utility.c:308]: Sure, but let's the speed penalty for Windows and so I would include the g_access() inside the Windows branch and do the check of st_mode on Linux.

Comment (jenshannoschwalm) [reply src/common/utility.c:308]: Are we sure we don't have ACL in use? access() would serve them too, stat wouldn't. I always feel pretty uneasy about this stuff, don't care here about performance but would always prefer high level functions.

Comment (TurboGit) [reply src/common/utility.c:308]: Agreed, I forgot about ACL which are not handled by stat().

**Review by TurboGit — APPROVED** [review]
So good after all. Thanks !

**Pull request merged**